fix(common): handle JS floating point errors in percent pipe

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?
```
[x] Bugfix
```

## What is the current behavior?
We multiply the number by 100 in the percent pipe which can lead to floating point errors with float numbers.

Issue Number: #20136


## What is the new behavior?
We now shift the decimal separator without using math, which avoids the js floating point error.
We also remove trailing 0 when rounding if they are not required by the minimal number of fraction (see added tests).


## Does this PR introduce a breaking change?
```
[x] No
```

ocombe committed:
fix(common): handle JS floating point errors in percent pipe

Fixes #20136

mary-poppins commented:
You can preview 6506c7c at https://pr20329-6506c7c.ngbuilds.io/.

vicb reviewed (changes requested):
I don't understand the issue we are trying to solve here.

ocombe commented:
@vicb WE are trying to get around this: http://0.30000000000000004.com/

vicb commented:
- Why are we trying to fix JS here and nowhere else ?,
- This is a breaking change,
- This change lacks docs,
- Why do we have a generic expensive function while we are only trying to shift the coma position ?,
- where is the code from ? (missing credits/license),
- where are the tests ?,
- are we sure that users expect this behavior rather than the regular JS behavior ?

ocombe commented:
https://jsperf.com/fix-js-floating-point-errors


> Why are we trying to fix JS here and nowhere else ?,
This is the only place where we change the number value (by a multiplication or something else)

> This is a breaking change,
No, this is a fix because it removes an unexpected behavior that no one wants to get

> This change lacks docs,
Added more comments on the function

> Why do we have a generic expensive function while we are only trying to shift the coma position ?,
Updated the function to be less generic since we only use it here

> where is the code from ? (missing credits/license),
Added link to the source in the comments of the function

> where are the tests ?
In packages/common/test/pipes/number_pipe_spec.ts, see the files changed by the commit

> are we sure that users expect this behavior rather than the regular JS behavior ?
We have a bug report about this and I tend to agree with it. Do you really expect to get `12.3456740000 %` when you use the percent pipe like this `{{ 0.12345674 | percent : '0.0-10' }}` ?

mary-poppins commented:
You can preview 58be008 at https://pr20329-58be008.ngbuilds.io/.

mary-poppins commented:
You can preview 1c547ba at https://pr20329-1c547ba.ngbuilds.io/.

mary-poppins commented:
You can preview 87a3980 at https://pr20329-87a3980.ngbuilds.io/.

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -43,11 +43,11 @@ export function formatNumber(
     return res;
   } else {
-    num = value;
+    num = value as number;
```
- ` as number` should not be required (as it is inferred) ?
- `let num: number` would be better IMO

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -367,6 +389,13 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
   }
+
+  // remove decimal trailing 0
+  let i = digits.length;
```
does it work with `1e100` ?
why do we have those 0 in the first place ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -273,6 +269,17 @@ function parseNumber(numStr: string): ParsedNumber {
     integerLen = numStr.length;
   }
 
+  // for the percent pipe, we multiply the number by 100
```
`parseNumber()` should parse a number.
Could you please add an extra function `toPercent(ParsedNumber): ParsedNumber`

ocombe replied:
parseNumber parses a number now, but I didn't add a new toPercent fonction because we don't have a ParsedNumber at this point

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -273,6 +269,17 @@ function parseNumber(numStr: string): ParsedNumber {
 
+  // for the percent pipe, we multiply the number by 100
+  if (style === NumberFormatStyle.Percent) {
+    const decimalLen = numStr.length - integerLen;
```
This line is far from being obvious to me by looking at the code.
Could you please add comments on the code above, rename the vars to more meaningful names.

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
   for (; fractionLen < Math.max(0, fractionSize); fractionLen++) digits.push(0);
 
-
+  let trailingZeros = !!fractionSize;
```
Be explicit. It is as fast and easier to read.

vicb replied:
Should it be `fractionLen` ? (if yes update the tests before fixing)

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
     d = d + carry;
     digits[i] = d % 10;
+    if (trailingZeros) {
+      // remove trailing 0 if necessary (e.g. 15.52000% --> 15.52%)
```
why `%` ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
+        trailingZeros = false;
+      }
+    }
     return Math.floor(d / 10);
```
Please replace `Math.floor(d / 10)` with `return d >= 10 ? 1 : 0; // Math.floor(d / 10)` (should be way faster)

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
   // Do any carrying, e.g. a digit was rounded up to 10
   const carry = digits.reduceRight(function(carry, d, i, digits) {
     d = d + carry;
     digits[i] = d % 10;
```
replace `d % 10;` with `d < 10 ? d : d - 10 // d % 10`

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
     digits[i] = d % 10;
+    if (trailingZeros) {
+      // remove trailing 0 if necessary (e.g. 15.52000% --> 15.52%)
+      if (digits[i] === 0 && i >= parsedNumber.integerLen + minFrac) {
```
Compute `parsedNumber.integerLen + minFrac` outside of the loop and give a meaningful name

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +363,19 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
   // Pad out with zeros to get the required fraction length
   for (; fractionLen < Math.max(0, fractionSize); fractionLen++) digits.push(0);
```
`digits.push(...new Array(Math.max(0. fractionSize)).fill(0));`

ocombe replied:
it doesn't work because it doesn't test the value of `fractionLen` compared to `Math.max(0, fractionSize)`

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -250,10 +245,11 @@ interface ParsedNumber {
 
 /**
- * Parse a number (as a string)
+ * Parse a number
```
Change to `Parses a number.`

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +364,20 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
 
-
+  let fixTrailingZeros = fractionSize !== 0;
+  const minLen = minFrac + parsedNumber.integerLen;
```
explain what this is and why ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -356,11 +364,20 @@ function roundNumber(parsedNumber: ParsedNumber, minFrac: number, maxFrac: numbe
   for (; fractionLen < Math.max(0, fractionSize); fractionLen++) digits.push(0);
 
-
+  let fixTrailingZeros = fractionSize !== 0;
```
`dropTrailingZeros` probably makes more sense. What are we "fix"ing ?

vicb commented on packages/common/src/i18n/format_number.ts:
```diff
@@ -249,11 +248,35 @@ interface ParsedNumber {
+
 /**
- * Parse a number (as a string)
+ * Parse a number.
```
PARSES

vicb commented:
LGTM once the CI is green. Great fix.
Please create an issue on AngularJS referencing this one.
